storage/shard.go

```diff
@@ -523,105 +569,118 @@ func (s *dbShard) tickAndExpire(
-			if len(expired) >= expireBatchLength {
-				// Purge when reaching max batch size to avoid large array growth
-				// and ensure smooth rate of elements being returned to pools.
-				// This method does not run using a lock so this is safe to
-				// perform inline.
-				s.purgeExpiredSeries(expired)
-				for i := range expired {
-					expired[i] = nil
+			if err == series.ErrSeriesAllDatapointsExpired {
+				expired = append(expired, entry)
+				r.expiredSeries++
+			} else {
+				r.activeSeries++
+				if err != nil {
+					r.errors++
 				}
-				expired = expired[:0]
 			}
-		} else {
-			r.activeSeries++
-			if err != nil {
-				r.errors++
+			r.activeBlocks += result.ActiveBlocks
+			r.openBlocks += result.OpenBlocks
+			r.wiredBlocks += result.WiredBlocks
+			r.unwiredBlocks += result.UnwiredBlocks
+			r.madeExpiredBlocks += result.MadeExpiredBlocks
+			r.madeUnwiredBlocks += result.MadeUnwiredBlocks
+			r.mergedOutOfOrderBlocks += result.MergedOutOfOrderBlocks
+			i++
+		}
+
+		// Purge any series requiring purging.
+		if len(expired) > 0 {
+			s.purgeExpiredSeries(expired)
+			for i := range expired {
+				expired[i] = nil
 			}
+			expired = expired[:0]
 		}
-		r.activeBlocks += result.ActiveBlocks
-		r.openBlocks += result.OpenBlocks
-		r.wiredBlocks += result.WiredBlocks
-		r.unwiredBlocks += result.UnwiredBlocks
-		r.madeExpiredBlocks += result.MadeExpiredBlocks
-		r.madeUnwiredBlocks += result.MadeUnwiredBlocks
-		r.mergedOutOfOrderBlocks += result.MergedOutOfOrderBlocks
-		i++
+
 		// Continue
 		return true
 	})
 
-	if len(expired) > 0 {
-		// Purge any series that still haven't been purged yet
-		s.purgeExpiredSeries(expired)
-	}
-
 	if terminatedTickingDueToClosing {
 		return tickResult{}, errShardClosingTickTerminated
 	}
 
 	return r, nil
 }
 
-func (s *dbShard) purgeExpiredSeries(expired []series.DatabaseSeries) {
+// NB(prateek): purgeExpiredSeries requires that all entries passed to it have at least one reader/writer,
+// i.e. have a readWriteCount of at least 1.
+// Currently, this function is only called by the lambda inside `tickAndExpire`'s `forEachShardEntryBatch`
+// call. This satisfies the contract of all entries it operating upon being guaranteed to have a
+// readerWriterEntryCount of at least 1, by virtue of the implementation of `forEachShardEntryBatch`.
+func (s *dbShard) purgeExpiredSeries(expiredEntries []*dbShardEntry) {
 	// Remove all expired series from lookup and list.
 	s.Lock()
-	for _, series := range expired {
+	for _, entry := range expiredEntries {
+		series := entry.series
 		hash := series.ID().Hash()
 		elem, exists := s.lookup[hash]
 		if !exists {
 			continue
 		}
-		entry := elem.Value.(*dbShardEntry)
+
+		count := entry.readerWriterCount()
+		// The contract requires all entries to have count >= 1.
+		if count < 1 {
+			s.logger.Errorf(`observed series [%s] with readerWriterCount [%d]. This violates
+				expected guarantees in the code.`, series.ID().String(), count)
```

Comment:
Hm this will actually print an extra end line in the logs and extra spacing in front of word "expected".

Better to just split with concat perhaps?

Reply:
Sure thing. Reworked to do the concat with the `Errorf` format string in the next patch.